libraries/abstractions/common_io/test/test_iot_uart.c

```diff
@@ -104,566 +145,647 @@ TEST_SETUP( TEST_IOT_UART )
+    iot_uart_set_callback( xUartHandle, prvReadWriteCallback, NULL );
 
-        lWrite = iot_uart_read_async( xOpen, cpBuffer, testIotUART_WRITE_BUFFER_LENGTH );
-        TEST_ASSERT_EQUAL( IOT_UART_SUCCESS, lWrite );
+    lWrite = iot_uart_write_async( xUartHandle, cpBuffer, testIotUART_BUFFER_LENGTH);
+    TEST_ASSERT_EQUAL( IOT_UART_SUCCESS, lWrite );
 
-        lIoctl = iot_uart_ioctl( xOpen, eGetRxNoOfbytes, &lTransferAmount );
-        TEST_ASSERT_EQUAL( IOT_UART_SUCCESS, lIoctl );
-    }
+    xCallbackReturn = xSemaphoreTake( xtestIotUARTSemaphore, testIotUART_DEFAULT_SEMPAHORE_DELAY );
+    TEST_ASSERT_EQUAL( pdTRUE, xCallbackReturn );
+
+    lRead = iot_uart_read_async( xUartHandle, cpBufferRead, testIotUART_BUFFER_LENGTH );
+    TEST_ASSERT_EQUAL( IOT_UART_SUCCESS, lRead );
 
-    lClose = iot_uart_close( xOpen );
+    xCallbackReturn = xSemaphoreTake( ( SemaphoreHandle_t ) &xtestIotUARTCompleted, testIotUART_DEFAULT_SEMPAHORE_DELAY );
+    TEST_ASSERT_EQUAL( pdTRUE, xCallbackReturn );
+
+    lClose = iot_uart_close( xUartHandle );
     TEST_ASSERT_EQUAL( IOT_UART_SUCCESS, lClose );
+
 }
 /*-----------------------------------------------------------*/
 
+/*-----------------------------------------------------------*/
 /**
- * @brief Test Function to cancel an asynchronous operation in flight.
- */
+ * @brief Test Function to test iot_uart_cancel
+ *
+*-----------------------------------------------------------*/
 TEST( TEST_IOT_UART, AFQP_IotUARTCancel )
 {
-    uint8_t cpBuffer[ testBufferSize ] = { 0 };
+    IotUARTHandle_t xUartHandle;
+    int32_t lIoctl, lWrite, lClose, lCancel;
 
-    IotUARTHandle_t xOpen;
-    int32_t lWrite, lCancel, lClose;
-    uint8_t ucPort = ustestIotUartPort;
-    BaseType_t xCallbackReturn;
+    uint8_t cSmallBuf[ 2 ] = { 'H', 'I' };
+    uint8_t cpBuffer[ testIotUARTBUFFERSIZE ] = { 0 };
+    uint8_t uSmallBuflen = sizeof ( cSmallBuf ) / sizeof ( uint8_t );
 
-    xOpen = iot_uart_open( ucPort );
-    TEST_ASSERT_NOT_EQUAL( NULL, xOpen );
+    strncpy( (char*)cpBuffer, testIotUART_BUFFER_STRING, testIotUART_BUFFER_LENGTH );
 
-    if( TEST_PROTECT() )
-    {
-        iot_uart_set_callback( xOpen, prvReadWriteCallback, NULL );
+    xUartHandle = iot_uart_open( ustestIotUartPort );
+    TEST_ASSERT_NOT_EQUAL( NULL, xUartHandle );
 
-        lWrite = iot_uart_write_async( xOpen, cpBuffer, testIotUART_WRITE_BUFFER_LENGTH );
-        TEST_ASSERT_EQUAL( IOT_UART_SUCCESS, lWrite );
+    iot_uart_set_callback( xUartHandle, prvCancelCallback, NULL );
 
-        lCancel = iot_uart_cancel( xOpen );
-        TEST_ASSERT_EQUAL( IOT_UART_SUCCESS, lCancel );
+    lWrite = iot_uart_write_async( xUartHandle, cpBuffer, testIotUART_BUFFER_LENGTH );
+    TEST_ASSERT_EQUAL( IOT_UART_SUCCESS, lWrite );
 
-        /* Wait to make sure operation was really canceled. */
-        xCallbackReturn = xSemaphoreTake( ( SemaphoreHandle_t ) &xtestIotUARTCompleted, testIotUART_DEFAULT_SEMPAHORE_DELAY );
```

Comment:
revert this, as semaphore will never be taken if cancel works

Reply:
Reverted the change, and using prvReadWriteCallback() instead. To ensure that callback is not getting called.